flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/service/context/SessionContext.java

```diff
@@ -235,40 +236,79 @@ public static SessionContext create(
 
         final ModuleManager moduleManager = new ModuleManager();
 
-        final EnvironmentSettings settings =
-                EnvironmentSettings.newInstance().withConfiguration(configuration).build();
-
-        CatalogManager catalogManager =
-                CatalogManager.newBuilder()
-                        // Currently, the classloader is only used by DataTypeFactory.
-                        .classLoader(userClassLoader)
-                        .config(configuration)
-                        .defaultCatalog(
-                                settings.getBuiltInCatalogName(),
-                                new GenericInMemoryCatalog(
-                                        settings.getBuiltInCatalogName(),
-                                        settings.getBuiltInDatabaseName()))
-                        .build();
+        final CatalogManager catalogManager =
+                buildCatalogManager(configuration, userClassLoader, environment);
 
         final FunctionCatalog functionCatalog =
                 new FunctionCatalog(configuration, resourceManager, catalogManager, moduleManager);
         SessionState sessionState =
                 new SessionState(catalogManager, moduleManager, resourceManager, functionCatalog);
 
-        return new SessionContext(
-                defaultContext,
-                sessionId,
-                endpointVersion,
-                configuration,
-                userClassLoader,
-                sessionState,
-                new OperationManager(operationExecutorService));
+        // --------------------------------------------------------------------------------------------------------------
+        // Build session context and return
+        // --------------------------------------------------------------------------------------------------------------
+
+        SessionContext sessionContext =
+                new SessionContext(
+                        defaultContext,
+                        sessionId,
+                        environment.getSessionEndpointVersion(),
+                        configuration,
+                        userClassLoader,
+                        sessionState,
+                        new OperationManager(operationExecutorService));
+
+        // filter the default catalog out to avoid exception caused by repeated registration
+        String currentCatalog = sessionContext.sessionState.catalogManager.getCurrentCatalog();
+        environment
+                .getRegisteredCatalogs()
+                .forEach(
+                        (catalogName, catalog) -> {
+                            if (!catalogName.equals(currentCatalog)) {
+                                sessionContext.registerCatalog(catalogName, catalog);
+                            }
+                        });
```

Comment:
Can we move this into the `buildCatalogManager`? I think this belongs to the initialization of the CatalogManager.

Reply:
Improved.